### DISCUSSION TITLE
Why setPWM works and manual config doesn't work?

### DISCUSSION THREAD

**Discussion by SuppaTenko** [stm32duino/Arduino_Core_STM32#1922] [Q&A] [answered]
Good day.
So, I'm using STM32G031G8U6 and I want high frequency PWM signal on Timer 14, Ch.1 - pin PB1. I need around 300...500 kHz to use really small inductor and p-fet to make inverting boost DC-DC for VFD project.
I clock MCU from internal OSC via PLL x8, so my core is 64MHz and my AHB and APB domains are also 64MHz. But situation is the same for default 16MHz clocks.
When i use simplified configuration - it works fine - i have nice and steady 82% duty and smooth -27 volts output.
`PWMTimer->setPWM(channel, pin, 500000, 82);`

But when i use something like this - output just stucks high.

```
PWMTimer->setMode(channel, TIMER_OUTPUT_COMPARE_PWM1, pin);
PWMTimer->setPrescaleFactor(1);
PWMTimer->setOverflow(128, TICK_FORMAT);
PWMTimer->setCaptureCompare(channel, 104, TICK_COMPARE_FORMAT);
PWMTimer->attachInterrupt(Update_IT_callback);
PWMTimer->attachInterrupt(channel, Compare_IT_callback);
```

This way it works up to 90kHz or something, and if i try to increase frequency - timer stucks.
Just in case, i checked all parameters via corresponding methods: getOverflow, getCount, getPrescaleFactor etc. And everything seems to be OK.
So, where's the error?  What am i doing wrong?

**Comment by fpistm** [comment]
Hii,

pay attention that PB1 has several possibilities:
https://github.com/stm32duino/Arduino_Core_STM32/blob/72524db53346fedaa8dd98122feb18bd97149be6/variants/STM32G0xx/G031G(4-6-8)U_G041G(6-8)U/PeripheralPins.c#L101C3-L103

So to used PB1 with TIM14 ch1 you have to use the pin number: `PB1_ALT2`.

**Reply by fpistm** [reply, marked as answer]
I've made some test with STM32G0316-DISCO and all configuration works as expected:

```C++
/* Comment to use all in one config */
#define FULL_CONFIG
#define pin  PB1_ALT2

void setup()
{
  Serial.begin(9600);
  // Automatically retrieve TIM instance and channel associated to pin
  // This is used to be compatible with all STM32 series automatically.
  TIM_TypeDef *Instance = (TIM_TypeDef *)pinmap_peripheral(digitalPinToPinName(pin), PinMap_PWM);
  uint32_t channel = STM_PIN_CHANNEL(pinmap_function(digitalPinToPinName(pin), PinMap_PWM));


  // Instantiate HardwareTimer object. Thanks to 'new' instantiation, HardwareTimer is not destructed when setup() function is finished.
  HardwareTimer *MyTim = new HardwareTimer(Instance);
  Serial.printf("Timer clk freq is: %u\n", MyTim->getTimerClkFreq());

  // Configure and start PWM
#ifndef FULL_CONFIG
  MyTim->setPWM(channel, pin, 500000, 82); // 500 kHz, 82% dutycycle
#else
  MyTim->setMode(channel, TIMER_OUTPUT_COMPARE_PWM1, pin);
  MyTim->setPrescaleFactor(1); // Useful for setOverflow with TICK_FORMAT
  // Timer clock freq is 64000000 Hz
  // So 128 Tick to get 2µs
  MyTim->setOverflow(128, TICK_FORMAT);
  // or
  // MyTim->setOverflow(2, MICROSEC_FORMAT); // 2 microseconds
  // or
  // MyTim->setOverflow(500000, HERTZ_FORMAT); // 500 kHz
  MyTim->setCaptureCompare(channel, 82, PERCENT_COMPARE_FORMAT); // 82%
  MyTim->resume();
#endif
}
```
![image](https://user-images.githubusercontent.com/20641798/213467024-e5ee7092-f37a-4a32-b8a6-aca47577483e.png)
